delivery_send_to_shipper_at_operation/tests/test_send_to_shipper.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2021 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+import types
+
+from odoo.tests.common import SavepointCase
+
+
+def send_shipping(self, pickings):
+    # Fake method bound dynamically to the carriers (monkey-patch)
+    # to get a delivery costs working without configuring a provider etc
+    return [{"exact_price": 10.0, "tracking_number": "TEST"}]
+
+
+class TestDeliverySendToShipper(SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
+        cls.wh = cls.env.ref("stock.warehouse0")
+        cls.wh.delivery_steps = "pick_pack_ship"
+        cls.stock_location = cls.env.ref("stock.stock_location_stock")
+        # cls.pack_location = cls.wh.wh_pack_stock_loc_id
+        cls.ship_location = cls.wh.wh_output_stock_loc_id
+        cls.customer_location = cls.env.ref("stock.stock_location_customers")
+        # put product in stock
+        cls.product = cls.env.ref("product.product_delivery_01")
+        cls.env["stock.quant"]._update_available_quantity(
+            cls.product, cls.stock_location, 10.0,
+        )
+        # create carriers
+        cls.delivery_fee = cls.env.ref("delivery.product_product_delivery")
+        cls.carrier_on_ship = cls.env["delivery.carrier"].create(
+            {
+                "name": "Carrier On Ship",
+                "product_id": cls.delivery_fee.id,
+                "integration_level": "rate_and_ship",
+                "send_delivery_notice_on": "ship",
+                "invoice_policy": "real",
+            }
+        )
+        cls.carrier_on_ship.__class__.send_shipping = types.MethodType(
+            send_shipping, cls.carrier_on_ship
+        )
+        cls.carrier_on_pack = cls.env["delivery.carrier"].create(
+            {
+                "name": "Carrier On Pack",
+                "product_id": cls.delivery_fee.id,
+                "integration_level": "rate_and_ship",
+                "send_delivery_notice_on": "custom",
+                "send_delivery_notice_picking_type_ids": [
+                    (6, 0, cls.wh.pack_type_id.ids)
+                ],
+                "invoice_policy": "real",
+            }
+        )
+        cls.carrier_on_pack.__class__.send_shipping = types.MethodType(
```

Comment:
why not mock or self.patch? this won't rollback you hack :thinking:

Reply:
On another topic I tested mock+side_effect to test the call while keeping the underlying method executed, I catched one drawback by doing so: we lose the user's context (and I was calling the mocked method with a specific context key). By monkey-patching with `types` there was no issue (context was updated as expected), so I keep this idea here, even if I don't rely on the context in this case.

Reply:
I switched to `mock` here